### ISSUE TITLE
Unable to open universal variable file on nfs

### ISSUE THREAD

**Issue by neolectron** [fish-shell/fish-shell#3362]
**fish version installed** : 2.3.1

**OS used**: osX el capitan

## Reproduction steps
0. Install fish as non-root user (via brew or .app file or even by rebuilding)
1. Have your session on a nfs (v3) filesystem.
2. Type any command.
3. Get the error.

## Results
```console
me@elcaptain ~/folder> ls
file1 file2
Unable to open universal variable file '/nfs/2015/m/me/.config/fish/fishd.10ddb1bddc7c': Operation not supported
me@elcaptain ~/folder> cd
Unable to open universal variable file '/nfs/2015/m/me/.config/fish/fishd.10ddb1bddc7c': Operation not supported
me@elcaptain ~> 
```
I noticed that i can't have the error while i'm recording an asciinema... dunno why.

I tried deleting those file, etc etc... :/

**Comment by zanchey** [comment]
I use fish on an NFSv3 server without too much trouble, so I wonder if there is a difference in our NFS server setups. `nfsstat -m` might help. Here's the output of mine:

```
nfsstat -m
/net/services.ucc.gu.uwa.edu.au/space/away/home from services.ucc.gu.uwa.edu.au:/space/away/home
  -- Original mount options:
     General mount flags: 0x500018 nodev,nosuid,automounted,nobrowse
     NFS parameters:
     File system locations:
       /space/away/home @ services.ucc.gu.uwa.edu.au (2405:3c00:5200:100::5,130.95.13.5)
  -- Current mount parameters:
     General mount flags: 0x4500018 nodev,nosuid,automounted,nobrowse multilabel
     NFS parameters: vers=3,tcp,port=2049,nomntudp,hard,nointr,noresvport,negnamecache,callumnt,locks,quota,rsize=32768,wsize=32768,readahead=16,dsize=32768,nordirplus,nodumbtimr,timeo=10,maxgroups=16,acregmin=5,acregmax=60,acdirmin=5,acdirmax=60,nomutejukebox,nonfc,sec=sys
     File system locations:
       /space/away/home @ services.ucc.gu.uwa.edu.au (2405:3c00:5200:100::5,130.95.13.5)
     Current location: 0x1 0 0 0:
       /space/away/home @ services.ucc.gu.uwa.edu.au (2405:3c00:5200:100::5)
     Status flags: 0x0
```

**Comment by zanchey** [comment]
(`locks` is what I'm wondering about)

**Comment by krader1961** [comment]
The problem is that fish only checks for `errno == EOPNOTSUP`. On MacOS (and probably other BSD drivatives) there is also `ENOTSUP` which is a distinct errno value. On Linux `EOPNOTSUP` is just an alias for `ENOTSUP`. See man 3 errno` this [discussion](https://cygwin.com/ml/libc-hacker/1999-01/msg00313.html) and this [message](https://mail-index.netbsd.org/tech-kern/2011/12/13/msg012185.html). We should check both symbols.

**Comment by zanchey** [comment]
The manual page on my OS X says that it returns `EOPNOTSUP` but doesn't mention `ENOTSUP`, although of course that may not match reality.

**Comment by floam** [comment]
I don't even have a "errno" on my OSX. 

It's mentioned in `man 2 intro`, though.

```
     45 ENOTSUP Not supported.  The attempted operation is not supported for
             the type of object referenced.
```

So is EOPNOTSUPP in `man 2 open`:
```man
     [EOPNOTSUPP]       O_SHLOCK or O_EXLOCK is specified, but the underlying
                        filesystem does not support locking.

     [EOPNOTSUPP]       An attempt is made to open a socket (not currently
                        implemented).
```

**Comment by floam** [comment]
`open -h errno` and pick sys/errno.h:

```
/* ipc/network software -- argument errors */
#define	ENOTSOCK	38		/* Socket operation on non-socket */
#define	EDESTADDRREQ	39		/* Destination address required */
#define	EMSGSIZE	40		/* Message too long */
#define	EPROTOTYPE	41		/* Protocol wrong type for socket */
#define	ENOPROTOOPT	42		/* Protocol not available */
#define	EPROTONOSUPPORT	43		/* Protocol not supported */
#if __DARWIN_C_LEVEL >= __DARWIN_C_FULL
#define	ESOCKTNOSUPPORT	44		/* Socket type not supported */
#endif
#define ENOTSUP		45		/* Operation not supported */
#if !__DARWIN_UNIX03 && !defined(KERNEL)
/*
 * This is the same for binary and source copmpatability, unless compiling
 * the kernel itself, or compiling __DARWIN_UNIX03; if compiling for the
 * kernel, the correct value will be returned.  If compiling non-POSIX
 * source, the kernel return value will be converted by a stub in libc, and
 * if compiling source with __DARWIN_UNIX03, the conversion in libc is not
 * done, and the caller gets the expected (discrete) value.
 */
#define	EOPNOTSUPP	 ENOTSUP	/* Operation not supported on socket */
#endif /* !__DARWIN_UNIX03 && !KERNEL */

#if __DARWIN_C_LEVEL >= __DARWIN_C_FULL
#define	EPFNOSUPPORT	46		/* Protocol family not supported */
#endif
#define	EAFNOSUPPORT	47		/* Address family not supported by protocol family */
#define	EADDRINUSE	48		/* Address already in use */
#define	EADDRNOTAVAIL	49		/* Can't assign requested address */
```

**Comment by krader1961** [comment]
@neolectron, Please see if this [change](https://github.com/krader1961/fish-shell/commit/37b8f584bdd5710e2d34faab46e7390a743848b8) fixes this failure on your system. It's a trivial change so you could also just s/EOPNOTSUPP/ENOTSUP/ in *src/env_universal_common.cpp*. It builds on macOS, Linux, and the three TravisCI environments we have set up for testing.

@zanchey and @floam, Yeah, I noticed that too but there's only one place we issue that specific error and the only plausible reason is that flag `O_EXLOCK` isn't supported. So I'm simply assuming that somewhere along the line macOS was changed to return `ENOTSUP` (the POSIX mandated symbol) rather than `EOPNOTSUP` (a legacy symbol apparently not sanctioned in any relevant standards).

**Comment by floam** [comment]
> ENOTSUP (the POSIX mandated symbol) rather than EOPNOTSUP (a legacy symbol apparently not sanctioned in any relevant standards).

Indeed, although OS X only turns off the alias if you compile under UNIX2003 conformance mode (we don't, fish gets `_DARWIN_C_FULL` with Apple's various nice extensions and occasional compatibility bits.)

So look like EOPNOTSUP should always work, on my 10.12 SDK at least. Should autoconf be checking that?

**Comment by krader1961** [comment]
@floam, I don't know. My guess based on our code and the observed behavior is that for whatever reason fish is actually getting ENOTSUP for some reason.

Also, I don't even see the point of using the O_EXLOCK flag in that open(2) call. The code will work just fine without it given that it falls back to trying flock(2). Which should also fail in this situation (i.e., a NFS mount without locking enabled) so I don't understand the thinking behind the current logic. Too, it shouldn't even be needed since the uvar file is supposedly unique for each machine (something we've talked about changing).

**Comment by floam** [comment]
> reason fish is actually getting ENOTSUP for some reason.

ENOTSUP and EOPNOTSUP (where the latter is defined), are actually the same, though, right?

**Comment by krader1961** [comment]
> ENOTSUP and EOPNOTSUP (where the latter is defined), are actually the same, though, right?

Supposedly but the convoluted `#if` blocks and associated comments in that C header make that hard to know with any certainty. And the observed behavior suggests they're not on the OPs system.

**Comment by krader1961** [comment]
It's a trivial to test that hypothesis so let's wait for feedback from the OP before we spend more time discussing this.

**Comment by floam** [comment]
I was mistaken about us not compiling with  __DARWIN_UNIX03 defined. I was confusing it with `__DARWIN_SUF_UNIX03` -- which is defined if `__DARWIN_UNIX03` and ` __DARWIN_ONLY_UNIX_CONFORMANCE`.

We will always have `__DARWIN_UNIX03`, (unless building with a 10.5 deployment target) so that makes more sense.

**Comment by krader1961** [comment]
FYI, Just to semi-answer a question I posed earlier the current logic (involving the use of `O_EXLOCK`) was introduced a little over two years ago by change 3b4794ae94157408de4cac33a50522be230feee0.

**Comment by krader1961** [comment]
@neolectron, Can you please test that substituting `ENOTSUP` for `EOPNOTSUPP` in *src/env_universal_common.cpp* fixes your problem? I'm repeating this because there have been multiple updates to this issue since my previous question of this nature and I want to ensure you understand we are waiting on you to verify our hypothesis and fix are correct.

**Comment by neolectron** [comment]
YES !!!!!
@krader1961 you fixed it with [this commit](https://github.com/krader1961/fish-shell/commit/37b8f584bdd5710e2d34faab46e7390a743848b8) 
thank you man 👍

**Comment by PeterDaveHello** [comment]
I still have the same problem on FreeBSD 11 with fish v2.5.0, anyone else? Thanks.

**Comment by zanchey** [comment]
@PeterDaveHello, if you're still experiencing this could you open a new issue with the details?

**Comment by PeterDaveHello** [comment]
@zanchey I'd like to! What kind of info do I need to provide please? Would OS version and fish version be enough? Thanks!

**Comment by krader1961** [comment]
@PeterDaveHello, Please open an issue with your OS details, fish version, home dir NFS mount options, and anything else you feel is relevant. We're probably going to also require you to build fish from source with custom patches to provide more details about the failure.
